src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -699,9 +890,15 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
         let is_non_exhaustive = is_privately_empty || is_declared_nonexhaustive;
 
         if missing_ctors.is_empty() && !is_non_exhaustive {
-            all_ctors.into_iter().map(|c| {
-                is_useful_specialized(cx, matrix, v, c.clone(), pcx.ty, witness)
-            }).find(|result| result.is_useful()).unwrap_or(NotUseful)
+            if value_constructors {
+                // If we've successfully matched every value
+                // of the type, then we're done.
+                NotUseful
```

Comment:
Why is this necessary (cc @nikomatsakis).

Reply:
See [the previous comment for my motivation](https://github.com/rust-lang/rust/pull/50912#discussion_r191266058).